src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -2091,14 +2091,28 @@ private BoundExpression MakeSizeOfMultiplication(BoundExpression numericOperand,
                 {
                     case SpecialType.System_Int32:
                         // add operator can take int32 and extend to 64bit if necessary
+                        // however in a case of checked operation, the operation is treated as unsigned with overflow ( add.ovf.un , sub.ovf.un )
+                        // the IL spec is a bit vague whether JIT should sign or zero extend the shorter operand in such case
+                        // and there could be inconsistencies in implementation or bugs.
+                        // As a result, in checked contexts, we will force sign-extending cast to be sure
+                        if (isChecked)
+                        {
+                            var constVal = numericOperand.ConstantValue;
+                            if (constVal == null || constVal.Int32Value < 0)
+                            {
+                                destinationType = SpecialType.System_IntPtr;
+                            }
+                        }
                         break;
                     case SpecialType.System_UInt32:
-                        // add operator treats operands as signed and will sign-extend on x64
-                        // to prevent sign-extending, convert the operand to unsigned native int.
-                        var constVal = numericOperand.ConstantValue;
-                        if (constVal == null || constVal.UInt32Value > int.MaxValue)
                         {
-                            destinationType = SpecialType.System_UIntPtr;
+                            // add operator treats operands as signed and will sign-extend on x64
+                            // to prevent sign-extending, convert the operand to unsigned native int.
+                            var constVal = numericOperand.ConstantValue;
```

Comment:
Why aren't we qualifying with `isChecked` here?

Reply:
C# compiler always assumed that shorter operand of mixed native/int operations is sign-extended regardless if operation is checked or not. It looks like sometimes some JITs indeed do that.
So we will continue forcing zero-extend cast here to be sure. This code did not really change.

Reply:
Why this is not symmetrical to the case above is that instruction set is not symmetrical either. There is `add`, `add.ovf` which are clearly sign-extending, there is also `add.ovf.un` which is at this point "depends on JIT/bugs, etc" and there is no `add.un` which would unambiguously zero-extend if it existed.  
With this change we will no longer depend on implicit behavior of `add.ovf.un` and `sub.ovf.un`

Reply:
Basically, if we had `add.un` we would be checking `isChecked` here.